### PR TITLE
Better missing value handle

### PR DESCRIPTION
**Pull request by guolinke**
Refer to #744 

add a parameter ```zero_as_missing```, True will use zeros to represent missing, otherwise will use `na` .

### CODE REVIEW

**Comment by guolinke** [comment]
@Laurae2 can you help to test these new behaviors ?

**Comment by Laurae2** [comment]
I take Bosch as example, with 500 iterations (for speed - MinGW 7.1) and reported best result.

Found 3 issues:

* Weird behavior when using a dense matrix (learn worse than random)
* Random Forest mode very slow (affects master branch)
* `best_iter` broken (always returns -1)

Install:

```r
devtools::install_github("Microsoft/LightGBM/R-package@2e82123") # Master
devtools::install_github("Microsoft/LightGBM/R-package@f76097e") # New NA handling
```

| Format |  LightGBM | AUC | Best Iter | Time |
| --- | --- | ---: | ---: | ---: |
| Dense | 2e82123 | 0.7133752 | 178 | 102279.974 |
| Dense | f76097e + FALSE | 0.6103862 | 36 | 172003.070 |
| Dense | f76097e + TRUE | 0.6103862 | 36 | 168656.152 |
| Sparse | 2e82123 | 0.7141376 | 169 | 102441.846 |
| Sparse | f76097e + FALSE | 0.7080622 | 356 | 107360.703 |
| Sparse | f76097e + TRUE | 0.7080622 | 356 | 103327.922 |

Something must be wrong in my script to test, do you have any idea?

```r
setwd("E:/datasets")
sparse <- TRUE
rf <- FALSE
zero_as_missing <- TRUE
if (rf == FALSE) {
  params <- list(num_threads = 40,
                 learning_rate = 0.05,
                 max_depth = 6,
                 num_leaves = 63,
                 max_bin = 255,
                 zero_as_missing = zero_as_missing)
} else {
  params <- list(num_threads = 40,
                 learning_rate = 1,
                 max_depth = -1,
                 num_leaves = 4097,
                 max_bin = 255,
                 zero_as_missing = zero_as_missing,
                 boosting_type = "rf",
                 bagging_freq = 1,
                 bagging_fraction = 0.632,
                 feature_fraction = ceiling(sqrt(970)) / 970)
}

library(data.table)
library(Matrix)
library(lightgbm)
library(R.utils)

data <- fread(file = "bosch_data.csv")


# Do xgboost / LightGBM

# When dense:
# > sum(data == 0, na.rm = TRUE)
# [1] 43574349
# > sum(is.na(data))
# [1] 929125166

# Split
if (sparse == TRUE) {
  library(recommenderlab)
  gc()
  train_1 <- dropNA(as.matrix(data[1:1000000, 1:969]))
  train_2 <- data[1:1000000, 970]$Response
  gc()
  test_1 <- dropNA(as.matrix(data[1000001:1183747, 1:969]))
  test_2 <- data[1000001:1183747, 970]$Response
  gc()
} else {
  gc()
  train_1 <- as.matrix(data[1:1000000, 1:969])
  train_2 <- data[1:1000000, 970]$Response
  gc()
  test_1 <- as.matrix(data[1000001:1183747, 1:969])
  test_2 <- data[1000001:1183747, 970]$Response
  gc()
}


# For LightGBM
train  <- lgb.Dataset(data = train_1, label = train_2)
test <- lgb.Dataset(data = test_1, label = test_2, reference=train)
train$construct()
test$construct()

gc()
Laurae::timer_func_print({temp_model <- lgb.train(params = params,
                                                  data = train,
                                                  nrounds = 500,
                                                  valids = list(test = test),
                                                  objective = "binary",
                                                  metric = "auc",
                                                  verbose = 2)})

# temp_model$best_iter
perf <- as.numeric(rbindlist(temp_model$record_evals$test$auc))
max(perf)
which.max(perf)
```

**Comment by guolinke** [comment]
@Laurae2 The best_iter will be always -1 if not using early stopping .

**Comment by guolinke** [comment]
@Laurae2 
```zero_as_missing``` should be passed when construct Dataset, if you calling ```construct()``` explicitly.

**Comment by Laurae2** [comment]
@guolinke I will re-run the small benchmarks on 7a359cb and come back with results soon

**Comment by Laurae2** [comment]
Includes now time for constructing dataset.

Install:

```r
devtools::install_github("Microsoft/LightGBM/R-package@2e82123") # Master
devtools::install_github("Microsoft/LightGBM/R-package@7a359cb") # New NA handling
```

GBT:

| Format | LightGBM | AUC | Best Iter | Time |
| --- | --- | ---: | ---: | ---: |
| Dense | 2e82123 | 0.7135208 | 159 | 207051.223 |
| Dense | 7a359cb + FALSE | 0.7119135 | 248 | 255843.066 |
| Dense | 7a359cb + TRUE | 0.6542308 | 10 | 157692.426 |
| Sparse | 2e82123 | 0.7130316 | 273 | 149938.056 |
| Sparse | 7a359cb + FALSE | 0.7097036 | 367 | 142425.563 |
| Sparse | 7a359cb + TRUE | 0.6542308 | 10 | 121133.742 |

RF: 

| Format | LightGBM | AUC | Time |
| --- | --- | ---: | ---: |
| Dense | 2e82123 | 0.6891714 | 156343.007 |
| Dense | 7a359cb + FALSE | 0.6959813 | 122706.562 |
| Dense | 7a359cb + TRUE | 0.5851799 | 160185.570 |
| Sparse | 2e82123 | 0.6891611 | 133761.146 |
| Sparse | 7a359cb + FALSE | 0.6885361 | 133258.710 |
| Sparse | 7a359cb + TRUE | 0.5855287 | 101993.678 |

Code for reproducibility:

```r
setwd("E:/datasets")
sparse <- TRUE # input sparse mode on/off
rf <- FALSE # Random Forest mode on/off
zero_as_missing <- FALSE # handle zero as missing?
zero_support <- TRUE # when using GitHub branch with zero_as_missing support
if (rf == FALSE) {
  params <- list(num_threads = 40,
                 learning_rate = 0.05,
                 max_depth = 6,
                 num_leaves = 63,
                 max_bin = 255)
} else {
  params <- list(num_threads = 40,
                 learning_rate = 1,
                 max_depth = -1,
                 num_leaves = 1023,
                 max_bin = 255,
                 boosting_type = "rf",
                 bagging_freq = 1,
                 bagging_fraction = 0.632,
                 feature_fraction = ceiling(sqrt(970)) / 970)
}
if (zero_support == TRUE) {
  params$zero_as_missing <- zero_as_missing
}

library(data.table)
library(Matrix)
library(R.utils)

data <- fread(file = "bosch_data.csv")


# Do xgboost / LightGBM

# When dense:
# > sum(data == 0, na.rm = TRUE)
# [1] 43574349
# > sum(is.na(data))
# [1] 929125166

# Split
if (sparse == TRUE) {
  library(recommenderlab)
  gc()
  train_1 <- dropNA(as.matrix(data[1:1000000, 1:969]))
  train_2 <- data[1:1000000, 970]$Response
  gc()
  test_1 <- dropNA(as.matrix(data[1000001:1183747, 1:969]))
  test_2 <- data[1000001:1183747, 970]$Response
  gc()
} else {
  gc()
  train_1 <- as.matrix(data[1:1000000, 1:969])
  train_2 <- data[1:1000000, 970]$Response
  gc()
  test_1 <- as.matrix(data[1000001:1183747, 1:969])
  test_2 <- data[1000001:1183747, 970]$Response
  gc()
}


# For LightGBM
# lgb.unloader(wipe = TRUE)
library(lightgbm)
train  <- lgb.Dataset(data = train_1, label = train_2)
test <- lgb.Dataset(data = test_1, label = test_2, reference=train)
# train$construct()
# test$construct()

gc()
Laurae::timer_func_print({temp_model <- lgb.train(params = params,
                                                  data = train,
                                                  nrounds = ifelse(rf == TRUE, 200, 500),
                                                  valids = list(test = test),
                                                  objective = "binary",
                                                  metric = "auc",
                                                  verbose = 2)})

perf <- as.numeric(rbindlist(temp_model$record_evals$test$auc))
max(perf)
which.max(perf)
```

**Comment by guolinke** [comment]
@Laurae2 
It is strange. The result with zero_as_missing=True should be same as the old solutions.

**Comment by Laurae2** [comment]
@guolinke Yes, this is strange. When I run my test on #744, I get 1 AUC with the correct number of bins (2 bins per feature).

```r
> model <- lgb.train(list(objective = "regression",
+                         metric = "auc",
+                         min_data = 1,
+                         learning_rate = 1,
+                         zero_as_missing = FALSE),
+                    dtrain,
+                    2,
+                    valids,
+                    early_stopping_rounds = 2)
[LightGBM] [Info] Total Bins 20
[LightGBM] [Info] Number of data: 10000, number of used features: 10
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=16 and max_depth=10
[1]:	test's auc:1 
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=1 and max_depth=1
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements.
[2]:	test's auc:1
```

image

**Comment by guolinke** [comment]
@Laurae2 
run you run the bosch data with following settings, and provide the accuracy and model file? 
1. 10 iterations with num_leaves=2, "master" branch 
2. 10 iterations with num_leaves=2, "missing" branch

**Comment by Laurae2** [comment]
Here are the files with 2 leaves and 10 iterations:

## Master branch

Model: [model_dense_master_10.txt](https://github.com/Microsoft/LightGBM/files/1185630/model_dense_master_10.txt)

Accuracy:

```r
> data.frame(1 - perf)
   X1...perf
1  0.9942802
2  0.9942802
3  0.9942802
4  0.9942802
5  0.9942802
6  0.9942802
7  0.9942802
8  0.9942802
9  0.9942802
10 0.9942802
```

## Missing branch

Model: [model_dense_missing_10.txt](https://github.com/Microsoft/LightGBM/files/1185633/model_dense_missing_10.txt)


Accuracy:

```r
> data.frame(1 - perf)
   X1...perf
1  0.9942802
2  0.9942802
3  0.9942802
4  0.9942802
5  0.9942802
6  0.9942802
7  0.9942802
8  0.9942802
9  0.9942802
10 0.9942802
```

**Comment by guolinke** [comment]
@Laurae2 
I fixed another bug. It should work now.

**Comment by Laurae2** [comment]
Seems fixed.

| Format | LightGBM | AUC | Best Iter | Time |
| --- | --- | ---: | ---: | ---: |
| Dense | 2e82123 | 0.7135208 | 159 | 207051.223 |
| Dense | 5a47feb + FALSE | 0.7119135 | 248 | 246036.848 |
| Dense | 5a47feb + TRUE | 0.7133752 | 178 | 192914.956 |

BTW, this is strange on Bosch 0 = NA is doing better. 0 has a meaning in that dataset (sensor reading in a path, there are several thousand paths).
